--- index.js
+++ index.js
@@ -10,13 +10,9 @@
 // 'extra' object is for decorating the error a bit more
 const promiseSpawn = (cmd, args, opts, extra = {}) => {
   const cwd = opts.cwd || process.cwd()
-  const isRoot = process.getuid && process.getuid() === 0
-  const { uid, gid } = isRoot ? inferOwner.sync(cwd) : {}
   return promiseSpawnUid(cmd, args, {
     ...opts,
-    cwd,
-    uid,
-    gid
+    cwd
   }, extra)
 }
 
